src/app/credExplorer/state.js

```diff
@@ -0,0 +1,247 @@
+  loadGraphWithAdapters: (repo: Repo) => Promise<GraphWithAdapters>;
+  pagerank: (
+    Graph,
+    EdgeEvaluator,
+    PagerankOptions
+  ) => Promise<PagerankNodeDecomposition>;
+
+  constructor(
+    getState: () => AppState,
+    setState: (AppState) => void,
+    loadGraphWithAdapters: (repo: Repo) => Promise<GraphWithAdapters>,
+    pagerank: (
+      Graph,
+      EdgeEvaluator,
+      PagerankOptions
+    ) => Promise<PagerankNodeDecomposition>
+  ) {
+    this.getState = getState;
+    this.setState = setState;
+    this.loadGraphWithAdapters = loadGraphWithAdapters;
+    this.pagerank = pagerank;
+  }
+
+  _maybeInitialize(state: Uninitialized): AppState {
+    const {repo, edgeEvaluator} = state;
+    if (repo != null && edgeEvaluator != null) {
+      const substate = {type: "READY_TO_LOAD_GRAPH", loading: "NOT_LOADING"};
+      return {type: "INITIALIZED", repo, edgeEvaluator, substate};
+    } else {
+      return state;
+    }
+  }
+
+  setRepo(repo: Repo) {
+    const state = this.getState();
+    switch (state.type) {
+      case "UNINITIALIZED": {
+        const newState = this._maybeInitialize({...state, repo});
+        this.setState(newState);
+        break;
+      }
+      case "INITIALIZED": {
+        const substate = {type: "READY_TO_LOAD_GRAPH", loading: "NOT_LOADING"};
+        const newState = {...state, repo, substate};
+        this.setState(newState);
+        break;
+      }
+      default: {
+        throw new Error((state.type: empty));
+      }
+    }
+  }
+
+  setEdgeEvaluator(edgeEvaluator: EdgeEvaluator) {
+    const state = this.getState();
+    switch (state.type) {
+      case "UNINITIALIZED": {
+        const newState = this._maybeInitialize({...state, edgeEvaluator});
+        this.setState(newState);
+        break;
+      }
+      case "INITIALIZED": {
+        const newState = {...state, edgeEvaluator};
+        this.setState(newState);
+        break;
+      }
+      default: {
+        throw new Error((state.type: empty));
+      }
+    }
+  }
+
+  async loadGraph() {
+    const state = this.getState();
+    if (
+      state.type !== "INITIALIZED" ||
+      state.substate.type !== "READY_TO_LOAD_GRAPH"
+    ) {
+      throw new Error("Tried to loadGraph in incorrect state");
+    }
+    const {repo, substate} = state;
+    const loadingState = {
+      ...state,
+      substate: {...substate, loading: "LOADING"},
+    };
+    this.setState(loadingState);
+    let newState: ?AppState;
+    try {
+      const graphWithAdapters = await this.loadGraphWithAdapters(repo);
+      newState = {
+        ...state,
+        substate: {
+          type: "READY_TO_RUN_PAGERANK",
+          graphWithAdapters,
+          loading: "NOT_LOADING",
+        },
+      };
+    } catch (e) {
+      console.error(e);
+      newState = {...state, substate: {...substate, loading: "FAILED"}};
+    }
+    if (deepEqual(this.getState(), loadingState)) {
+      this.setState(NullUtil.get(newState));
```

Comment:
You can omit this `NullUtil.get`; Flow correctly infers that `newState`
cannot be `null`.